lib/vsdata.cpp

```diff
@@ -1270,8 +1270,8 @@ void VisualizationSceneScalarData::Init()
    arrow_type = arrow_scaling_type = 0;
    scaling = 0;
    drawaxes = colorbar = 0;
-   auto_ref_max = 16;
-   auto_ref_max_surf_elem = 20000;
+   auto_ref_max = 32;
+   auto_ref_max_surf_elem = 5000000;
```

Comment:
I don't think we have a need to increase `auto_ref_max` -- 16 is more than sufficient for any practical case. For extreme cases it can be increased manually with the keys.

Also, I think increasing `auto_ref_max_surf_elem` to $5\times 10^6$ is too excessive. An increase by `25x` from $2\times 10^4$ to $5\times 10^5$ is more reasonable, however, even that may cause undesired, noticeable slowdowns, especially for lower order meshes (e.g. linear meshes) -- note that the refinement is applied independent of degree, so that curvature from the non-linear terms in $Q_1$ meshes/fields can be captured.

Another option is to keep the current value and instead issue a warning in the terminal (we may want to add a feature where terminal messages are shown for a brief moment in the main window and then fade away) when the auto-selected refinement factor is less than the max of the orders of the mesh and the order of the field.

Reply:
I think what we want ideally is that these are set to the maximum values for what a "reasonable" mesh may need, and then for GLVis to select those the actual auto-refinement values based on the order of the mesh (and solution)

Reply:
Yes, maximum of some reasonable value and a value based on the grid function + mesh order 👍 , but it could be capped by some higher number maybe? To give the user a chance to change the value and not burn off the computer right away 🔥 😄

Reply:
For me, "reasonable" auto-selected refinement factor is mainly driven by the time it takes to display the initial mesh. If possible, displaying the initial mesh should not take longer than ~0.1-0.3 sec on relatively new hardware. If we roughly say that the time to display some surface is proportional to to the point evaluations we need to perform, then we need to choose an upper bound for `auto_ref_max_surf_elem` which gives the desired time limit. Of course, achieving this limit may not be possible when the number of the drawn surface elements is not too big.

In the methods `VisualizationSceneSolution::GetAutoRefineFactor()` and `VisualizationSceneSolution3d::GetAutoRefineFactor()`, the number of point evaluations is computed as `ne*(ref+1)*(ref+1)` where `ne` is the number of surface elements in the mesh (i.e. number of elements in 2D meshes, embedded in 2D or 3D, or the number of boundary elements for 3D meshes). So the question is: how many point evaluations can we do in the allowed ~0.1-0.3 sec -- we can debate what exact value to aim for in this range. The evaluation speed per point will depend to some extend on the orders of the mesh and the field, so to make things more concrete, let's say we should measure the evaluation times for $Q_2$ meshes/fields.

Does this sound like a "reasonable" approach to determine the upper limit `auto_ref_max_surf_elem`?

Reply:
Regardless of how we do the auto-selection of refinement level, users will need to be aware that for big meshes they may not get any element sub-divisions.

Reply:
> For example `auto_ref_max` should be at least `max{ element_order, gf_order}`, maybe scaled by a factor of 2?

This sounds good but I'd still allow up to 16 refinement factor for lower orders, so formula like `auto_ref_max = max(16, 2*mesh_order, 2*field_order)`seems better -- it should ok for 20th order, 1-element the spiral mesh. 😁

Reply:
Speed is important, but for me avoiding confusion is also important, and I think we should consider sacrificing some of the former for the latter.

The issue is that sometimes we refine automatically and sometimes we don't. If for example we never refined by default the users will quickly learn to do it themselves (as in VisIt). If on the other hand we always refine perfectly, then they might not even know about pressing 'o'. 

A warning message is a good idea. Mentioning it in CHANGELOG will also help.

But ideally, if we are going for the "automatic" case, we need some minimal refinement to at least indicate that there is a curvature.

Reply:
If loading and displaying a high-order mesh + solution using refinement factor 1 takes a long time, say 30 sec, then I don't think we want to use anything more than 1 as the auto-selected refinement factor.

Because of this, I think we want users to be aware that the refinement factor is auto-selected and will be smaller for bigger meshes down to the limiting case of refinement factor of 1. I don's see a good way to avoid this.

Reply:
@psocratis and @dylan-copeland -- what are you thoughts on this?

Reply:
It sounds like automation is not generally practical, and that the user will need to be aware that the visualization is under-resolved, requiring their input to fully resolve. Even if we increase the hard-coded maximum values, the same problems will be encountered at some mesh size. A clear warning message seems to be the only solution, but we could also increase the maximum values, still to a conservative level ensuring speed. It is possible that a user would not notice a warning message printed to the console, so I wonder if it is easy to put a warning also in the visualization window (which could be removed with a key?). To summarize, I am in favor of (i) clear warnings to the console and maybe the visualization window; and (ii) a modest increase in maximums to ensure speed.

Reply:
There was some unclarity about the possible algorithm. What I was thinking of was 1) sum the order-based number of refinements over the mesh (`order_based`), 2) `autoref_max_surf_elem=min(max(order_based, 20k), 5M)`, 3) if `order_based > 5M`, show a warning.

Reply:
To add to ☝️

The [current algorithm](https://github.com/GLVis/glvis/blob/master/lib/vssolution.cpp#L942-L952) is essentially

```c++
int ne = mesh->GetNE(), ref = 1;
while (ref < 16 && ne*(ref+1)*(ref+1) <= 20000) { ref++; }
```

What I was thinking about is something like:

```c++
int ne = mesh->GetNE(), ref = 1;

if (ne*(order_ref+1)*(order_ref+1) < 2000000) // try to be accurate (within reason)
{
   ref = order_ref;
}
else // we can't be accurate, so issue a warning and be fast
{
   while (ref < 16 && ne*(ref+1)*(ref+1) <= 100000) { ref++; }
}
```

The exact numbers (2M and 100K) could be adjusted of course.

Reply:
That is in that spirit, but there is no minimum, so it would be even degrading some visualization, so I would keep that 20k there (something similar). And then it is not much logical to me to jump from 2M to 100K, I would expect it to be continuous 🤔 .

Reply:
I assume in (1) you want to use a uniform (we cannot actually draw non-uniform) subdivision factor based on the maximal order in the mesh and in the solution. So the result from (1) is `num_elems*max_order*max_order` -- this corresponds to candidate value `max_order` for the subdivision factor. Then from that, you push the result from (1) into the interval `[20k, 5M]`, if it is outside, and that is the result from (2); finally, based on (2), find the maximal subdivision factor, `sd`, such that `num_elems*sd*sd` does not exceed the number from (2).

So, compared to what we have now, instead of `num_elems*sd*sd <= 20k` you want to use an upper limit based on `max_order`, namely `num_elems*sd*sd <= min(max(num_elems*max_order*max_order, 20k), 5M)`. This basically says: "I'm willing to wait X seconds for order p=4, but for order p=2, I'm only willing to wait X/4 seconds." I'm not sure this is a good default. What makes more sense to me is to say "I'm willing to wait X seconds", independent of the order, and with X that is relatively small, e.g. 0.1-0.3; from there, let users adjust things if they are willing to wait more.

In the end, it is a matter of preference -- what do others prefer?

Reply:
My interpretation of @najlkin's suggestion can be a bit misleading: "I'm willing to wait X seconds for order p=4, but for order p=2, I'm only willing to wait X/4 seconds" should be more specific and add "on a mesh with the same `num_elem`". If we compare "p=4 on one mesh and p=2 on the refined mesh" then @najlkin's suggestion puts the same limit on time for both meshes -- roughly speaking the time limit depends on the number of dofs, in 2D, or the number of boundary dofs, in 3D.

Reply:
What I had in mind is different: "I'm willing to wait X seconds for accurate visualization, but if it can not be accurate I don't want to wait"

Reply:
@tzanio, One problem with your suggestion is that for small meshes you'll get `ref = order_ref` whereas now you can get `ref = 16`.

Reply:
> One problem with your suggestion is that for small meshes you'll get ref = order_ref whereas now you can get ref = 16.

If `order_ref` means "the visualization is accurate" then I think that's fine. We can of course also check for that and increase it if 16 works.

My main point was to try to be "accurate" however we define that in terms of the order. Right now, we are just trying to be fast.

Reply:
I don't think we want to treat "the visualization is accurate" as a Boolean in this context -- it is a discrete range with `ref = 1` being the least accurate, and larger `ref` being more accurate. The question is how to balance the accuracy levels with the speed (which we can assume is also proportional to resource usage).

To put it simply: the modification in your algorithm that I was thinking of, is to move the `while` loop after the `if else` statement. It is still not perfect but it is an improvement.

Reply:
Yes, I agree it should not be a boolean, so moving the while loop like this. It just does not make sense to me to do the jump from high resolution to low in the if statement, I though of the truncation in the sense: "We tried to visualize the grid function accurately, but we could not in reasonable time, so here is the best we could do reasonably."
Besides that, I would keep that 20k as minimum which costs nearly nothing and it keeps backward compatibility in terms of visualization quality.

Reply:
In @tzanio's suggestion, if the `while` loop is after the `if else` statement, the 20k bound we have now is replaced by 100k bound, so there will be no degradation of quality compared to now.

Reply:
There could be if `order_ref` will give a number lower than 20k, or no? 🤔 .

Reply:
The `order_ref` is just an initial value for `ref` before the `while` loop (when the `if` is true). So if `order_ref < 16 && ne*(order_ref+1)*(order_ref+1) <= 100000`, you will get more refinements than `order_ref` and no less refinements than before.

Reply:
The problem, for me, with @tzanio's suggestion (even when the `while` loop is moved after the `if else` statement) is that it can result in big jumps in speed for small change in the input. For example, if `ne*(order_ref+1)*(order_ref+1)` is close but less than 2M we get `ref=order_ref` and if we increase `ne` just a little to push `ne*(order_ref+1)*(order_ref+1)` beyond 2M, then can immediately go down to `ref=1`.

Reply:
Aha, like this, yes, in the modified version it would work, but I thought it is more logical to iterate to the same limiting number as in the criterion for `order_ref` to keep continuity and not jump down, as you say.

Reply:
What I was proposing would solve it (updated version to have the same numbers):
```
int ne = mesh->GetNE(), ref = 1;
int autoref_max = min(max(ne*(order_ref+1)*(order_ref+1), 100k), 2M)
while (ref < 16 && ne*(ref+1)*(ref+1) <= autoref_max) { ref++; }
```